src/data/repositories/ProgramLoggerD2Repository.ts

```diff
@@ -25,20 +27,45 @@ export class ProgramLoggerD2Repository implements LoggerRepository {
         this.messageId = dataElements.messageId;
         this.messageTypeId = dataElements.messageTypeId;
         this.organisationUnitId = organisationUnitId;
+        this.programStage = undefined;
+    }
+
+    async init(): Promise<void> {
+        this.programStage = await this.getProgramStage().toPromise();
     }
```

Comment:
`this.programStage = undefined;` raises a flag: imperative pattern. I see the problem, you have some async initialization which cannot be in the constructor. The typical solution for this would to be:

- Add "programStage" as argument to the constructor
- static async init(): Promise<void> { ... return new ProgramLoggerD2Repository(...) }

I think this will work here. Sometimes this pattern does not fit to the caller. Then we are forced to make this.programStage a property (cached) which returns a FutureData<D2ProgramStage> -> requires to refactor methods using it. Let's see if option (1) works.